lib/modules/platform/bitbucket-server/utils.ts

```diff
@@ -156,39 +158,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
   return invalidReviewers;
 }
 
+function generateUrlFromEndpoint(
+  defaultEndpoint: string,
+  opts: HostRule,
+  repository: string
+): string {
+  const url = new URL(defaultEndpoint);
+  const generatedUrl = git.getUrl({
+    protocol: url.protocol as GitProtocol,
+    auth: `${opts.username}:${opts.password}`,
+    host: `${url.host}${url.pathname}${
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+      url.pathname!.endsWith('/') ? '' : /* istanbul ignore next */ '/'
+    }scm`,
+    repository,
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
+
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption | undefined,
   info: BbsRestRepo,
   opts: HostRule
-): string {
-  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
-  if (!cloneUrl) {
-    // Http access might be disabled, try to find ssh url in this case
-    cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');
-  }
+): string | null {
+  if (gitUrl === undefined) {
```

Comment:
@viceice do the latest typing changes mean that this is now not possible - would be `null`?